0006-bingings.md

```diff
@@ -0,0 +1,591 @@
+* `header_hash` -- `h256`/`[u8; 32]`, hash of the header on Ethereum blockchain. We use the pointer to the slice of 32 bytes on guest memory;
+* `nonce` -- `u64`/`h64`/`[u8; 8]`, nonce that was used to find the correct hash, passed as `u64` without pointers;
+* `mix_hash` -- `h256`/`[u8; 32]`, special hash that avoid griefing attack. We use the pointer to the slice of 32 bytes on guest memory;
+* `difficulty` -- `u256`/`[u64; 4]`, the difficulty of mining the block. We use the pointer to the slice of 32 bytes on guest memory;
+
+###### Returns
+* `1` if the Ethash is valid;
+* `0` otherwise.
+
+###### Panics
+* If `block_number_ptr + 32` or `header_hash_ptr + 32` or `mix_hash_ptr + 32` or `difficulty_ptr + 32` point outside the memory or scratch buffer uses more memory than the limit with `MemoryAccessViolation`.
+
+###### Current bugs
+* `block_number` and `difficulty` are currently exposed as `u64` which are casted to `u256` which breaks Ethereum compatibility;
+* Currently, we also pass the length together with `header_hash_ptr` and `mix_hash_ptr` which is not necessary since
+we know their length.
+
+## Promises API
+
+```rust
+promise_create(account_id_len: u64,
+               account_id_ptr: u64,
+               method_name_len: u64,
+               method_name_ptr: u64,
+               arguments_len: u64,
+               arguments_ptr: u64,
+               amount_ptr: u64) -> u64
+```
+Creates a promise that will execute a method on account with given arguments and attaches the given amount.
+`amount_ptr` points to a slice of bytes representing `u128`.
+
+###### Panics
+* If `account_id_len + account_id_ptr` or `method_name_len + method_name_ptr` or `arguments_len + arguments_ptr`
+or `amount_ptr + 8` points outside the memory of the guest or host, with `MemoryAccessViolation`.
+
+###### Returns
+* Index of the new promise that uniquely identifies it within the current execution of the method.
+
+---
+
+```rust
+promise_then(promise_idx: u64,
+             account_id_len: u64,
+             account_id_ptr: u64,
+             method_name_len: u64,
+             method_name_ptr: u64,
+             arguments_len: u64,
+             arguments_ptr: u64,
+             amount_ptr: u64) -> u64            
+```
+Attaches the callback that is executed after promise pointed by `promise_idx` is complete.
+
+###### Panics
+* If `promise_idx` does not correspond to an existing promise panics with `InvalidPromiseIndex`.
+* If `account_id_len + account_id_ptr` or `method_name_len + method_name_ptr` or `arguments_len + arguments_ptr`
+or `amount_ptr + 8` points outside the memory of the guest or host, with `MemoryAccessViolation`.
+
+###### Returns
+* Index of the new promise that uniquely identifies it within the current execution of the method.
+
+---
+```rust
+promise_and(promise_id1: u64, promise_id2: u64) -> u64
```

Comment:
That can be changed to take a C style array instead of 2 indices

Reply:
Done